lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp

```diff
@@ -623,20 +623,103 @@ GetObjectFileFormat(llvm::Triple::ObjectFormatType obj_format_type) {
+/// return nothing if the swift symbols were stripped.
+///
+/// 2 - The file name itself.
+///
+/// We use those two sources of information to come up with an educated guess of
+/// what the swift module's name(s) is (are).
 static llvm::SmallVector<llvm::StringRef, 1>
-GetLikelySwiftImageNamesForModule(ModuleSP module) {
+GetPossibleSwiftModuleNames(ModuleSP module) {
   if (!module || !module->GetFileSpec())
     return {};
 
-  auto name =
+
+  auto filename =
       module->GetFileSpec().GetFileNameStrippingExtension().GetStringRef();
-  if (name == "libswiftCore")
-    name = "Swift";
-  if (name.startswith("libswift"))
-    name = name.drop_front(8);
-  if (name.startswith("lib"))
-    name = name.drop_front(3);
-  return {name};
+  // We know that "libswiftCore" contains the "Swift"  swift module.
+  if (filename == "libswiftCore")
+    filename = "Swift";
+  // We also know that "libswiftFoo" binaries we ship usually contain the "Foo"
+  // swift module.
+  if (filename.startswith("libswift"))
+    filename = filename.drop_front(8);
+  // "libFoo.dylib" is a common naming scheme for shipping dynamic libraries 
+  // with a swift module "Foo".
+  if (filename.startswith("lib"))
+    filename = filename.drop_front(3);
+
+  SymbolContextList symbol_context_list;
+  // A module descriptor is mangled as the module mangling followed by 'MXM'.
+  module->FindSymbolsMatchingRegExAndType(RegularExpression(".*MXM$"),
+                                          lldb::SymbolType::eSymbolTypeData,
+                                          symbol_context_list, true);
+
+  llvm::SmallVector<llvm::StringRef, 1> swift_module_names;
+  for (size_t i = 0; i < symbol_context_list.GetSize(); ++i) {
+    SymbolContext sc;
+    if (symbol_context_list.GetContextAtIndex(i, sc) && sc.symbol) {
+      auto mangled_name = sc.symbol->GetMangled().GetMangledName().GetStringRef();
+      auto module_name = GetModuleNameFromModuleDescriptor(mangled_name);
+      if (module_name.empty())
+        continue;
+
+      // Since almost every image will contain objc types, checking for that
+      // module name isn't a great filter.
+      if (!module_name.empty() && module_name != "__C" &&
+          module_name != "__C_Synthesized")
+        swift_module_names.emplace_back(module_name);
+    }
+  }
+
+  // If we didnt't find any module descriptors they most likely were stripped,
+  // the filename is the best that we can do.
+  if (swift_module_names.empty()) 
+    return {filename};
+
+  // If we found exactly one module descriptor, that's almost certainly the 
+  // canonical module descriptor, so just return that.
+  if (swift_module_names.size() == 1)
+    return swift_module_names;
+
+  // If the filename exactly matches one of the module descriptor's module name,
+  // we can be reasonably certain that's the canonical module descriptor.
+  if (std::find(swift_module_names.begin(), swift_module_names.end(),
```

Comment:
The only thing that worries me with the current implementation is the case where an image contains multiple canonical swift modules where one is the "main" module and matches the binary name exactly. 

For example, say we have the following canonical modules in the  `Foo`, `FooUtils`, `FooMaps`, `FooWhatever`, `SomeRandomModule` in the image, and the file name is `Foo`. This will discard those other modules.

On the other hand, if we don't implement this check, we'll have many more false positives with the swift modules the image references, so I'm not sure what's better.

Reply:
@adrian-prantl what do you think?